System/Process/CommunicationHandle.hsc

```diff
@@ -0,0 +1,371 @@
+          pfdStdOutput inheritWrite overlappedWrite
+      let ((ourPtr, ourMode), (theirPtr, _theirMode)) =
+            mbSwap ((pfdStdInput, ReadMode), (pfdStdOutput, WriteMode))
+      ourHANDLE  <- peek ourPtr
+      theirHANDLE <- peek theirPtr
+      -- With WinIO, we need to associate any handles we are going to use in
+      -- the current process before being able to use them.
+      return ()
+##  if defined (__IO_MANAGER_WINIO__)
+        <!> associateHandle' ourHANDLE
+##  endif
+      ourHandle <-
+##  if !defined (__IO_MANAGER_WINIO__)
+        rawFdToHandle ourMode =<< openHANDLE ourHANDLE
+##  else
+        -- NB: it's OK to call the following function even when we're not
+        -- using WinIO at runtime, so we don't use <!>.
+        rawHANDLEToHandle ourMode ourHANDLE
+##  endif
+      return $ (ourHandle, CommunicationHandle theirHANDLE)
+
+foreign import ccall "mkNamedPipe" c_mkNamedPipe ::
+    Ptr HANDLE -> Bool -> Bool -> Ptr HANDLE -> Bool -> Bool -> IO Bool
+
+##endif
+
+--------------------------------------------------------------------------------
+
+-- | A version of 'readCreateProcessWithExitCode' that communicates with the
+-- child process through a pair of 'CommunicationHandle's.
+--
+-- Example usage:
+--
+-- > readCreateProcessWithExitCodeCommunicationHandle
+-- >   (\(chTheyRead, chTheyWrite) -> proc "child-exe" [show chTheyRead, show chTheyWrite])
+-- >   (\ hWeRead -> hGetContents hWeRead)
+-- >   (\ hWeWrite -> hPut hWeWrite "xyz")
+--
+-- where @child-exe@ is a separate executable that is implemented as:
+--
+-- > main = do
+-- >   [chRead, chWrite] <- getArgs
+-- >   hRead  <- useCommunicationHandle $ read chRead
+-- >   hWrite <- useCommunicationHandle $ read chWrite
+-- >   input <- hGetContents hRead
+-- >   hPut hWrite $ someFn input
+-- >   hClose hWrite
+--
+-- @since 1.6.19.0
+readCreateProcessWithExitCodeCommunicationHandle
+  :: NFData a
+  => ((CommunicationHandle, CommunicationHandle) -> CreateProcess)
+    -- ^ Process to spawn, given a @(read, write)@ pair of
+    -- 'CommunicationHandle's that are inherited by the spawned process
+  -> (Handle -> IO a)
+    -- ^ read action
+  -> (Handle -> IO ())
+    -- ^ write action
+  -> IO (ExitCode, a)
+readCreateProcessWithExitCodeCommunicationHandle mkProg readAction writeAction = do
+  (chTheyRead, hWeWrite   ) <- createTheyReadWeWritePipe
+  (hWeRead   , chTheyWrite) <- createWeReadTheyWritePipe
+  let cp = mkProg (chTheyRead, chTheyWrite)
+  -- The following implementation parallels 'readCreateProcess'
+  withCreateProcess cp $ \ _ _ _ ph -> do
+    -- Close the parent's references to the 'CommunicationHandle's after they
+    -- have been inherited by the child (we don't want to keep pipe ends open).
+    closeCommunicationHandle chTheyWrite
+    closeCommunicationHandle chTheyRead
+
+    -- Fork off a thread that waits on the output.
+    output <- readAction hWeRead
+    withForkWait (evaluate $ rnf output) $ \ waitOut -> do
```

Comment:
`rnf` looks suspicious. What is it for?

Is it for forcing all lazy IO before closing `hWeRead`? If so then that's merely a sticking plaster, not a proper solution, because I can easily subvert the `NFData` instance by writing my own on a `newtype` wrapper. Therefore I would prefer not to impose the `NFData` constraint at all.

If it's for something else, could you add a comment to explain?

https://github.com/haskell/process/pull/308/files#diff-edde8bd0b4546b5a2c8aad3f7a290cec8d6c6cd698e7c5d707be3a2a847e446cR364

Reply:
This logic is copied from [`readCreateProcess`](https://github.com/haskell/process/blob/master/System/Process.hs#L520), except that that function uses `String` to communicate with the child process. I think we should not decide for the user of the API which type we use for inter-process communication (and potential users such as `GHC` certainly do not want to just use `String`), so I made the code parametric over this type (requiring an `NFData` instance for that type).

I agree that all this use of lazy IO makes me a bit uneasy, but I preferred to re-use the existing implementation that exists in process rather than doing it completely differently. I am not against taking a different approach that avoids lazy IO, but perhaps that can be the topic of another MR that updates the existing code as well? (I am not necessarily opposed to only changing this code path, but being consistent across the library in this regard seems more sensible to me.)

Reply:
Ah yes, I see. I agree it makes sense to have this new code be consistent with existing code.

In fact, can't we always use this API by setting `a ~ ()`? Instead of returning a complete value it seems like we can just act on the data we receive incrementally in `readAction`.

Reply:
I'm not sure about your suggestion to use `a ~ ()`. Perhaps consumers can modify their code if the API was changed in that way, but I preferred to keep the API the same as for `readCreateProcess`, just with different types than `String`. I hope that's OK.

Reply:
Yes, that's fine.

My comment about `a ~ ()` was that I think that no value of `a` other than `()` is actually *required* to use this API. You can simulate other desired behavior for other values of `a` with a carefully chosen `readAction`. But I agree with the point that consistency with existing functions is important.